python/ray/_private/metrics_agent.py

```diff
@@ -60,43 +66,109 @@ def name(self):
-                prometheus_exporter.Options(
-                    namespace="ray",
-                    port=metrics_export_port,
-                    address=metrics_export_address,
-                )
-            )
-        except Exception:
-            # TODO(SongGuyang): Catch the exception here because there is
-            # port conflict issue which brought from static port. We should
-            # remove this after we find better port resolution.
-            logger.exception(
-                "Failed to start prometheus stats exporter. Agent will stay "
-                "alive but disable the stats."
-            )
+        #
+        # Opencensus components to record metrics.
+        #
+
+        # Managing views to export metrics
+        # If the view manager is None, we disable all metrics export.
+        self.view_manager = view_manager
+        # A class that's used to record metrics
+        # emitted from the current process.
+        self.stats_recorder = stats_recorder
+        # A class to export metrics.
+        self.stats_exporter = stats_exporter
+
+        if self.stats_exporter is None:
+            # If the exporter is not given,
+            # we disable metrics collection.
             self.view_manager = None
         else:
             self.view_manager.register_exporter(stats_exporter)
 
-    def record_reporter_stats(self, records: List[Record]):
-        if not self.view_manager:
-            return
+        # Below fields are used to clean up view data when workers are dead.
+        # Note that Opencensus export metrics from
+        # `view_data.tag_value_aggregation_data_map[tag_vals]`, which means
+        # we can stop exporting metrics by deleting data in there.
+        # Note that tag_vals is something like (<WorkerId>, <IP>, <tags..>).
+        # Each worker metrics will have its own unique tag_vals because they
+        # always contain WorkerId.
+        # - When we proxy exports metrics, if they are from a worker, we store
+        #   the corresponding {worker_id -> tag vals} and last reported time to
+        #   `WorkerProxyExportState`.
+        # - We periodically go over all `WorkerProxyExportState`. If the last
+        #   reported time is bigger than the threashold, we treat
+        #   the worker as dead.
+        # - If the worker somehow reports metrics again, we starts reporting
+        #   again. But if it stops reporting for a long time, it will be
+        #   eventually marked as dead.
+
+        # {worker_id -> {view_name -> {tag_vals}}}
+        # Used to clean up data created from a worker of worker id.
+        self.worker_id_to_state = defaultdict(WorkerProxyExportState)
+        # After the timeout, the worker is marked as dead.
+        # The timeout is reset every time worker reports
+        # new metrics (it happens every 2 seconds by default).
+        # This value must be longer than the Prometheus
+        # scraping interval (10s by default in Ray).
```

Comment:
is it possible to enforce some checks on the constants here if breaking the contract would be bad?

Reply:
Unfortunately, it is impossible to get prometheus scraping interval... maybe we should document this instead. I will follow up.

Reply:
@ericl I doubled the timeout because the default prom scraping interval is 1 minute (it is 10 seconds in Ray config, but it is prone to error to set this 1 minute if users use external prom)